File: 0/Templates/VisualBasic/ProjectTemplates/SplitViewAppVBTemplate/Presentation/ShellViewModel.vb
Imports Intense.Presentation

Namespace Presentation
    Public Class ShellViewModel
        Inherits NotifyPropertyChanged

        Private _topItems As NavigationItemCollection = New NavigationItemCollection
        Private _selectedTopItem As NavigationItem
        Private _bottomItems As NavigationItemCollection = New NavigationItemCollection
        Private _selectedBottomItem As NavigationItem
        Private _isSplitViewPaneOpen As Boolean
        Private _toggleSplitViewPaneCommand As ICommand

        Public Sub New()
            _toggleSplitViewPaneCommand = New RelayCommand(Sub() IsSplitViewPaneOpen = Not IsSplitViewPaneOpen)

            ' open splitview pane in wide state
            IsSplitViewPaneOpen = IsWideState()
        End Sub

        Public ReadOnly Property ToggleSplitViewPaneCommand As ICommand
            Get
                Return _toggleSplitViewPaneCommand
            End Get
        End Property

        Public Property IsSplitViewPaneOpen As Boolean
            Get
                Return _isSplitViewPaneOpen
            End Get
            Set
                Me.Set(_isSplitViewPaneOpen, Value)
            End Set
        End Property

        Public Property SelectedTopItem As NavigationItem
            Get
                Return _selectedTopItem

            End Get
            Set
                If Me.Set(_selectedTopItem, Value) And Not Value Is Nothing Then
                    OnSelectedMenuItemChanged(True)
                End If
            End Set
        End Property

        Public Property SelectedBottomItem As NavigationItem
            Get
                Return _selectedBottomItem

            End Get
            Set
                If Me.Set(_selectedBottomItem, Value) And Not Value Is Nothing Then
                    OnSelectedMenuItemChanged(False)
                End If
            End Set
        End Property

        Public Property SelectedItem As NavigationItem
            Get
                If Not _selectedTopItem Is Nothing Then
                    Return _selectedTopItem
                End If
                Return _selectedBottomItem
            End Get
            Set
                SelectedTopItem = _topItems.FirstOrDefault(Function(m) m.Equals(Value))
                SelectedBottomItem = _bottomItems.FirstOrDefault(Function(m) m.Equals(Value))
            End Set
        End Property

        Public Property SelectedPageType As Type
            Get
                Dim item = SelectedItem
                If Not item Is Nothing Then
                    Return item.PageType
                End If
                Return Nothing
            End Get
            Set
                ' select associated menu item
                SelectedTopItem = _topItems.FirstOrDefault(Function(m) m.PageType.Equals(Value))
                SelectedBottomItem = _bottomItems.FirstOrDefault(Function(m) m.PageType.Equals(Value))
            End Set
        End Property

        Public ReadOnly Property TopItems As NavigationItemCollection
            Get
                Return _topItems
            End Get
        End Property

        Public ReadOnly Property BottomItems As NavigationItemCollection
            Get
                Return _bottomItems
            End Get
        End Property

        Private Sub OnSelectedMenuItemChanged(top As Boolean)
            If top Then
                SelectedBottomItem = Nothing
            Else
                SelectedTopItem = Nothing
            End If
            OnPropertyChanged("SelectedItem")
            OnPropertyChanged("SelectedPageType")

            ' auto-close split view pane (only when not in widestate)
            If Not IsWideState() Then
                IsSplitViewPaneOpen = False
            End If
        End Sub

        ' a helper determining whether we are in a wide window state
        ' mvvm purists probably don't appreciate this approach
        Private Function IsWideState() As Boolean
            Return Window.Current.Bounds.Width >= 1024
        End Function

    End Class
End Namespace

File: 0/Templates/CSharp/ProjectTemplates/SplitViewAppTemplate/Presentation/ShellViewModel.cs
using System;
using System.Collections.ObjectModel;
using System.Linq;
using System.Windows.Input;
using Windows.UI.Xaml;

namespace $safeprojectname$.Presentation
{
    public class ShellViewModel : NotifyPropertyChanged
    {
        private ObservableCollection<MenuItem> menuItems = new ObservableCollection<MenuItem>();
        private MenuItem selectedMenuItem;
        private ObservableCollection<MenuItem> bottomMenuItems = new ObservableCollection<MenuItem>();
        private MenuItem selectedBottomMenuItem;
        private bool isSplitViewPaneOpen;

        public ShellViewModel()
        {
            this.ToggleSplitViewPaneCommand = new Command(() => this.IsSplitViewPaneOpen = !this.IsSplitViewPaneOpen);

            // open splitview pane in wide state
            this.IsSplitViewPaneOpen = IsWideState();
        }

        public ICommand ToggleSplitViewPaneCommand { get; private set; }

        public bool IsSplitViewPaneOpen
        {
            get { return this.isSplitViewPaneOpen; }
            set { Set(ref this.isSplitViewPaneOpen, value); }
        }

        public MenuItem SelectedMenuItem
        {
            get { return this.selectedMenuItem; }
            set
            {
                if (Set(ref this.selectedMenuItem, value) && value != null) {
                    OnSelectedMenuItemChanged(true);
                }
            }
        }

        public MenuItem SelectedBottomMenuItem
        {
            get { return this.selectedBottomMenuItem; }
            set
            {
                if (Set(ref this.selectedBottomMenuItem, value) && value != null) {
                    OnSelectedMenuItemChanged(false);
                }
            }
        }

        public Type SelectedPageType
        {
            get
            {
                return (this.selectedMenuItem ?? this.selectedBottomMenuItem)?.PageType;
            }
            set
            {
                // select associated menu item
                this.SelectedMenuItem = this.menuItems.FirstOrDefault(m => m.PageType == value);
                this.SelectedBottomMenuItem = this.bottomMenuItems.FirstOrDefault(m => m.PageType == value);
            }
        }

        public ObservableCollection<MenuItem> MenuItems
        {
            get { return this.menuItems; }
        }

        public ObservableCollection<MenuItem> BottomMenuItems
        {
            get { return this.bottomMenuItems; }
        }

        private void OnSelectedMenuItemChanged(bool top)
        {
            if (top) {
                this.SelectedBottomMenuItem = null;
            }
            else {
                this.SelectedMenuItem = null;
            }
            OnPropertyChanged("SelectedPageType");

            // auto-close split view pane (only when not in widestate)
            if (!IsWideState()) {
                this.IsSplitViewPaneOpen = false;
            }
        }

        // a helper determining whether we are in a wide window state
        // mvvm purists probably don't appreciate this approach
        private bool IsWideState()
        {
            return Window.Current.Bounds.Width >= 1024;
        }
    }
}
